File: src/main/java/benchmarks/programs/thread_context_switch/ThreadContextSwitch.java
package benchmarks.programs.thread_context_switch;

import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;
import java.util.concurrent.Semaphore;

public class ThreadContextSwitch {

    static class ThreadFunction implements Runnable {
        private final Semaphore ownSemaphore;
        private final Semaphore nextSemaphore;

        public ThreadFunction(Semaphore ownSemaphore, Semaphore nextSemaphore) {
            this.ownSemaphore = ownSemaphore;
            this.nextSemaphore = nextSemaphore;
        }

        @Override
        public void run() {
            try {
                while (true) {
                    ownSemaphore.acquire();
                    nextSemaphore.release();
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }
    }

    public static int getNumberOfTests(String[] args) {
        if (args.length < 2) {
            System.out.println("Not enough arguments. Usage: <program> <number_of_tests>");
            System.exit(1);
        }

        int numOfTests = 0;
        try {
            numOfTests = Integer.parseInt(args[1]);
            if (numOfTests <= 0) {
                throw new IllegalArgumentException("Invalid input: Expected a positive integer.");
            }
        } catch (Exception e) {
            System.out.println("Invalid input: " + e.getMessage());
            System.exit(1);
        }

        return numOfTests;
    }

    public static void main(String[] args) {
        int numOfTests = getNumberOfTests(args);

        Semaphore semaphore1 = new Semaphore(0);
        Semaphore semaphore2 = new Semaphore(0);

        Thread thread1 = new Thread(new ThreadFunction(semaphore1, semaphore2));
        Thread thread2 = new Thread(new ThreadFunction(semaphore2, semaphore1));

        thread1.start();
        thread2.start();

        double totalSwitchTime = 0.0;
        long start, end;

        try (BufferedWriter writer = new BufferedWriter(new FileWriter("..\\..\\result_files\\thread_context_switch\\java_thread_context_switch_result.txt"))) {
            for (int i = 0; i < numOfTests + 1; i++) {
                semaphore1.release();

                start = System.nanoTime();
                semaphore2.acquire();
                end = System.nanoTime();

                double switchTime = (end - start) / 1_000.0;
                if (i != 0) {
                    totalSwitchTime += switchTime;
                    writer.write(i + " " + String.format("%.3f", switchTime));
                    writer.newLine();
                }
            }

            double averageSwitchTime = totalSwitchTime / numOfTests;
            writer.write(String.format("%.3f", averageSwitchTime));
        } catch (IOException | InterruptedException e) {
            System.out.println(e.getMessage());
        }

        thread1.interrupt();
        thread2.interrupt();
    }
}

File: src/main/java/benchmarks/programs/thread_context_switch/thread_context_switch.c
#include <windows.h>
#include <stdio.h>
#include <stdlib.h>

DWORD WINAPI ThreadFunction(LPVOID lpParam) {
    HANDLE* events = (HANDLE*)lpParam;
    HANDLE ownEvent = events[0];
    HANDLE nextEvent = events[1];

    while (1) {
        WaitForSingleObject(ownEvent, INFINITE);
        SetEvent(nextEvent);
    }

    return 0;
}

int getNumberOfTests(int argc, char* argv[]) {
    if (argc < 3) {
        printf("Not enough arguments. Usage: <program> <number_of_tests>\n");
        exit(1);
    }

    int numOfTests;
    if (sscanf(argv[2], "%d", &numOfTests) != 1 || numOfTests <= 0) {
        printf("Invalid input: Expected a positive integer.\n");
        exit(1);
    }

    return numOfTests;
}

int main(int argc, char* argv[]) {
    SYSTEM_INFO systemInfo;
    LARGE_INTEGER start, end, frequency;
    HANDLE hThread1, hThread2;
    HANDLE event1, event2;
    DWORD_PTR originalAffinityMask;

    GetSystemInfo(&systemInfo);
    int numProcessors = systemInfo.dwNumberOfProcessors;

    if (numProcessors < 2) {
        printf("This test requires at least 2 processors.\n");
        return 1;
    }

    int numOfTests = getNumberOfTests(argc, argv);

    event1 = CreateEvent(NULL, TRUE, FALSE, NULL);
    event2 = CreateEvent(NULL, TRUE, FALSE, NULL);

    HANDLE thread1Params[2] = { event1, event2 };
    HANDLE thread2Params[2] = { event2, event1 };

    hThread1 = CreateThread(NULL, 0, ThreadFunction, thread1Params, 0, NULL);
    hThread2 = CreateThread(NULL, 0, ThreadFunction, thread2Params, 0, NULL);

    originalAffinityMask = SetThreadAffinityMask(hThread1, 1);
    SetThreadAffinityMask(hThread2, 1);

    QueryPerformanceFrequency(&frequency);
    double totalSwitchTime = 0.0;

    FILE* file = fopen("..\\..\\result_files\\thread_context_switch\\c_thread_context_switch_result.txt", "w");
    if (file == NULL) {
        printf("Failed to open the file for writing.\n");
        CloseHandle(hThread1);
        CloseHandle(hThread2);
        CloseHandle(event1);
        CloseHandle(event2);
        return 1;
    }


    for (int i = 0; i < numOfTests + 1; i++) {
        SetEvent(event1);

        QueryPerformanceCounter(&start);
        WaitForSingleObject(event2, INFINITE);
        QueryPerformanceCounter(&end);

        ResetEvent(event1);
        ResetEvent(event2);

        double switchTime = ((double)(end.QuadPart - start.QuadPart) * 1e6) / frequency.QuadPart;
        if(i != 0 ){
            totalSwitchTime += switchTime;
            fprintf(file, "%d %.3f\n", i, switchTime);
        }
    }

    double averageSwitchTime = totalSwitchTime / numOfTests;
    fprintf(file, "%.3f\n", averageSwitchTime);

    fclose(file);

    SetThreadAffinityMask(hThread1, originalAffinityMask);
    TerminateThread(hThread1, 0);
    TerminateThread(hThread2, 0);
    CloseHandle(hThread1);
    CloseHandle(hThread2);
    CloseHandle(event1);
    CloseHandle(event2);

    return 0;
}
